src/bin/pgcopydb/pgsql_timeline.c

```diff
@@ -0,0 +1,370 @@
+		context->parsedOk = false;
+		return;
+	}
+
+	/* systemid (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	if (!stringToUInt64(value, &(context->system->identifier)))
+	{
+		log_error("Failed to parse system_identifier \"%s\"", value);
+		context->parsedOk = false;
+		return;
+	}
+
+	/* timeline (int4) */
+	value = PQgetvalue(result, 0, 1);
+	if (!stringToUInt32(value, &(context->system->timeline)))
+	{
+		log_error("Failed to parse timeline \"%s\"", value);
+		context->parsedOk = false;
+		return;
+	}
+
+	/* xlogpos (text) */
+	value = PQgetvalue(result, 0, 2);
+	strlcpy(context->system->xlogpos, value, PG_LSN_MAXLENGTH);
+
+	/* dbname (text) Database connected to or null */
+	if (!PQgetisnull(result, 0, 3))
+	{
+		value = PQgetvalue(result, 0, 3);
+		strlcpy(context->system->dbname, value, NAMEDATALEN);
+	}
+
+	context->parsedOk = true;
+}
+
+
+/*
+ * parseTimelineHistoryResult parses the result of the TIMELINE_HISTORY
+ * replication command.
+ */
+static void
+parseTimelineHistoryResult(void *ctx, PGresult *result)
+{
+	TimelineHistoryResult *context = (TimelineHistoryResult *) ctx;
+
+	if (PQnfields(result) != 2)
+	{
+		log_error("Query returned %d columns, expected 2", PQnfields(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) == 0)
+	{
+		log_sql("parseTimelineHistoryResult: query returned no rows");
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) != 1)
+	{
+		log_error("Query returned %d rows, expected 1", PQntuples(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	/* filename (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	strlcpy(context->filename, value, sizeof(context->filename));
+
+	/* content (bytea) */
+	value = PQgetvalue(result, 0, 1);
+	context->content = strdup(value);
+
+	if (context->content == NULL)
+	{
+		log_error("parseTimelineHistoryResult: Failed to allocate memory for "
+				  "timeline history file of %zu bytes",
+				  strlen(value));
+		context->parsedOk = false;
+		return;
+	}
+
+	context->parsedOk = true;
+}
+
+
+/*
+ * parseTimelineHistory parses the content of a timeline history file.
+ */
+bool
+parseTimelineHistory(const char *filename, const char *content,
+					 IdentifySystem *system, void *context)
+{
+	LinesBuffer lbuf = { 0 };
+
+	if (!splitLines(&lbuf, (char *) content))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	uint64_t prevend = InvalidXLogRecPtr;
+
+	TimelineHistoryEntry *entry = &system->currentTimeline;
+	int timelineCount = 0;
+
+	for (uint64_t lineNumber = 0; lineNumber < lbuf.count; lineNumber++)
+	{
+		char *ptr = lbuf.lines[lineNumber];
+
+		/* skip leading whitespace and check for # comment */
+		for (; *ptr; ptr++)
+		{
+			if (!isspace((unsigned char) *ptr))
+			{
+				break;
+			}
+		}
+
+		if (*ptr == '\0' || *ptr == '#')
+		{
+			continue;
+		}
+
+		log_trace("parseTimelineHistory line %lld is \"%s\"",
+				  (long long) lineNumber,
+				  lbuf.lines[lineNumber]);
```

Comment:
If we get to store the history file as-is on-disk then we should be able to remove that line entirely.

Reply:
Yeah that's why the placement of my comment about storing to disk was at this seemingly random place. I would like to store the file exactly as Postgres makes it available to us, only for debugging purposes, in case our parsing of the file were to fail.

Reply:
I see now what you've done, and I think that's good. When implementing debug traces that are not used in the normal flow of operations, or even in the code, I don't think separation of concerns is as imperative. Here, we want to log-to-disk the actual result from the Postgres protocol, before any kind of processing...

Reply:
After reading the following sentence in the last comment, I realized that I did not print a debug message when writing the tli history file.

> When implementing debug traces that are not used in the normal flow of operations, or even in the code, I don't think separation of concerns is as imperative.

I pushed a new commit 2eacb24 that adds a debug message similar to what we had prior to this PR.

Reply:
Only thinking about it now, sorry: what could make sense is to store the timeline data received from the protocol directly to disk, no parsing, and then have a function that reads the file one line at a time and parses the content etc.

This would be better at memory management, and also separation of concerns and internal API/headers.

What do you think?

Reply:
I actually write contents to disk before parsing it.

Currently `pgsql_identify_system` function does the following calls:

- parseTimelineHistoryResult parses PGResult into a TimelineHistoryResult
- writeTimelineHistoryFile writes the TimelineHistoryResult content that holds raw data to disk
- parseTimelineHistoryfile parses the TimelineHistoryResult and stores it in internal catalog

Reply:
See the other comments I have added, with more details. I think we have an opportunity to implement things in a way that `start_logical_decoding` does not need access to our SQLite catalogs.